plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/TopQueriesByLatencyService.java

```diff
@@ -0,0 +1,240 @@
+ * top N queries with high latency data for search queries
+ *
+ * @opensearch.internal
+ */
+public class TopQueriesByLatencyService extends QueryInsightsService<
+    SearchQueryLatencyRecord,
+    PriorityBlockingQueue<SearchQueryLatencyRecord>,
+    QueryInsightsExporter<SearchQueryLatencyRecord>> {
+    private static final Logger log = LogManager.getLogger(TopQueriesByLatencyService.class);
+
+    private static final TimeValue delay = TimeValue.ZERO;
+
+    private int topNSize = QueryInsightsSettings.DEFAULT_TOP_N_SIZE;
+
+    private TimeValue windowSize = TimeValue.timeValueSeconds(QueryInsightsSettings.DEFAULT_WINDOW_SIZE);
+
+    private final ClusterService clusterService;
+    private final Client client;
+
+    @Inject
+    public TopQueriesByLatencyService(ThreadPool threadPool, ClusterService clusterService, Client client) {
+        super(threadPool, new PriorityBlockingQueue<>(), null);
+        this.clusterService = clusterService;
+        this.client = client;
+    }
+
+    /**
+     * Ingest the query data into to the top N queries with latency store
+     *
+     * @param timestamp The timestamp of the query.
+     * @param searchType The manner at which the search operation is executed. see {@link SearchType}
+     * @param source The search source that was executed by the query.
+     * @param totalShards Total number of shards as part of the search query across all indices
+     * @param indices The indices involved in the search query
+     * @param propertyMap Extra attributes and information about a search query
+     * @param phaseLatencyMap Contains phase level latency information in a search query
+     * @param tookInNanos Total search request took time in nanoseconds
+     */
+    public void ingestQueryData(
+        final Long timestamp,
+        final SearchType searchType,
+        final String source,
+        final int totalShards,
+        final String[] indices,
+        final Map<String, Object> propertyMap,
+        final Map<String, Long> phaseLatencyMap,
+        final Long tookInNanos
+    ) {
+        if (timestamp <= 0) {
+            log.error(
+                String.format(
+                    Locale.ROOT,
+                    "Invalid timestamp %s when ingesting query data to compute top n queries with latency",
+                    timestamp
+                )
+            );
+            return;
+        }
+        if (totalShards <= 0) {
+            log.error(
+                String.format(
+                    Locale.ROOT,
+                    "Invalid totalShards %s when ingesting query data to compute top n queries with latency",
+                    totalShards
+                )
+            );
+            return;
+        }
+        this.threadPool.schedule(() -> {
+            super.ingestQueryData(
+                new SearchQueryLatencyRecord(timestamp, searchType, source, totalShards, indices, propertyMap, phaseLatencyMap, tookInNanos)
+            );
```

Comment:
How threadsafe is "store"?  If it is threadsafe, could you please add a comment in the code?

Does this need to happen in a different thread?